x/ibc/02-client/types/client_test.go

```diff
@@ -0,0 +1,56 @@
+package types_test
+
+import (
+	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+	"github.com/cosmos/cosmos-sdk/x/ibc/exported"
+	ibctesting "github.com/cosmos/cosmos-sdk/x/ibc/testing"
+)
+
+func (suite *TypesTestSuite) TestMarshalConsensusStateWithHeight() {
+	var (
+		cswh types.ConsensusStateWithHeight
+	)
+
+	testCases := []struct {
+		name     string
+		malleate func()
+	}{
+		{
+			"solo machine client", func() {
+				soloMachine := ibctesting.NewSolomachine(suite.T(), suite.chainA.Codec, "solomachine", "")
+				cswh = types.NewConsensusStateWithHeight(types.NewHeight(0, soloMachine.Sequence), soloMachine.ConsensusState())
+			},
+		},
+		{
+			"tendermint client", func() {
+				clientA, _ := suite.coordinator.SetupClients(suite.chainA, suite.chainB, exported.Tendermint)
+				clientState := suite.chainA.GetClientState(clientA)
+				consensusState, ok := suite.chainA.GetConsensusState(clientA, clientState.GetLatestHeight())
+				suite.Require().True(ok)
+
+				cswh = types.NewConsensusStateWithHeight(clientState.GetLatestHeight().(types.Height), consensusState)
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			tc.malleate()
+
+			cdc := suite.chainA.App.AppCodec()
+
+			// marshal message
+			bz, err := cdc.MarshalJSON(&cswh)
+			suite.Require().NoError(err)
+
+			// unmarshal message
+			newCswh := &types.ConsensusStateWithHeight{}
+			err = cdc.UnmarshalJSON(bz, newCswh)
+			suite.Require().NoError(err)
```

Comment:
I tried testing for equivalency but I couldn't get this to work. Not sure how to test if the old and new are equal

Reply:
Do they have `Equal`. Maybe you need to deference first?

Reply:
ah, no they don't have `Equal`. I suppose that should only be necessary if we do like `reflect.DeepEqual` in the code? (which doesn't happen for this struct). 

I tried various combinations of `reflect.DeepEqual` and `proto.Equal` with references and dereferenced values. I'm starting to think the check would be unnecessary though if the marshal/unmarshal doesn't throw an error. I think I should just assume the unmarshal/marshal functions work as expected. 

I mostly just wanted to ensure that not using a pointer for this type when embedded in genesis doesn't cause any issues

Reply:
Ahhh, if you don't need Equal then no need to add it, but there is no harm in adding the gogo-proto annotation.

> I'm starting to think the check would be unnecessary though if the marshal/unmarshal doesn't throw an error. I think I should just assume the unmarshal/marshal functions work as expected.

With JSON, I would _not_ count on this. You can get **no** errors even when unmarshaling _different_ types (e.g. https://play.golang.org/p/gDaN3uV_14p).

What's the diff when doing `suite.Require().Equal(a, b)`?

Reply:
```
Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -35,23 +35,3 @@
                	            	   }),
                	            	-  compat: (*types.anyCompat)({
                	            	-   aminoBz: ([]uint8) <nil>,
                	            	-   jsonBz: ([]uint8) (len=230) {
                	            	-    00000000  7b 22 40 74 79 70 65 22  3a 22 2f 69 62 63 2e 74  |{"@type":"/ibc.t|
                	            	-    00000010  65 6e 64 65 72 6d 69 6e  74 2e 43 6f 6e 73 65 6e  |endermint.Consen|
                	            	-    00000020  73 75 73 53 74 61 74 65  22 2c 22 74 69 6d 65 73  |susState","times|
                	            	-    00000030  74 61 6d 70 22 3a 22 32  30 32 30 2d 30 31 2d 30  |tamp":"2020-01-0|
                	            	-    00000040  32 54 30 30 3a 30 30 3a  30 30 5a 22 2c 22 72 6f  |2T00:00:00Z","ro|
                	            	-    00000050  6f 74 22 3a 7b 22 68 61  73 68 22 3a 22 64 66 32  |ot":{"hash":"df2|
                	            	-    00000060  44 67 48 67 72 2f 73 35  66 6c 6b 66 73 61 6b 50  |DgHgr/s5flkfsakP|
                	            	-    00000070  77 52 6d 6c 44 37 31 72  6b 30 4e 68 6e 4c 62 36  |wRmlD71rk0NhnLb6|
                	            	-    00000080  36 6f 35 57 4f 34 50 51  3d 22 7d 2c 22 6e 65 78  |6o5WO4PQ="},"nex|
                	            	-    00000090  74 5f 76 61 6c 69 64 61  74 6f 72 73 5f 68 61 73  |t_validators_has|
                	            	-    000000a0  68 22 3a 22 33 34 39 32  35 41 44 42 46 32 41 30  |h":"34925ADBF2A0|
                	            	-    000000b0  45 33 36 45 42 34 35 36  31 33 43 30 39 33 42 44  |E36EB45613C093BD|
                	            	-    000000c0  36 43 32 42 30 30 42 41  43 38 37 38 33 39 31 38  |6C2B00BAC8783918|
                	            	-    000000d0  41 43 44 43 36 39 39 37  30 34 41 41 37 37 45 33  |ACDC699704AA77E3|
                	            	-    000000e0  35 44 43 43 22 7d                                 |5DCC"}|
                	            	-   },
                	            	-   err: (error) <nil>
                	            	-  })
                	            	+  compat: (*types.anyCompat)(<nil>)
```

Reply:
Ohhh the `compat` is the diff. I think that's because the any isn't cached yet? @aaronc how do we get the `compat` to populate?

Reply:
can we clear the cached value maybe?